### PR TITLE
added auto-complete for tag command

### LINKED ISSUE #594: Add auto-completion to /tag
Slashcommands now have autocompletion support.

`/tag` is the perfect candidate for this and should support it.

Therefore, simply implement `onAutoComplete` which is already avaialble to `SlashCommandAdapter` and incorperate existing tags into it.

### PR DESCRIPTION
**Pull request by Taz03**
Added auto complete for tag command

image

closes #594

### CODE REVIEW

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 1]
```diff
@@ -62,4 +70,19 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
         }
         message.queue();
     }
+
+    @Override
+    public void onAutoComplete(CommandAutoCompleteInteractionEvent event) {
+        if (event.getName().equals(COMMAND_NAME)
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:76]: This is useless

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 2]
```diff
@@ -62,4 +70,19 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
+                && event.getFocusedOption().getName().equals(ID_OPTION)) {
+            List<Command.Choice> choices = tagSystem.getAllIds()
+                .stream()
+                .filter(id -> id.startsWith(event.getFocusedOption().getValue()))
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:80]: I'd instead refactor the StringDistance method to get *good* autocompletion instead of a simple startsWith

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 3]
```diff
@@ -21,24 +25,28 @@
 public final class TagCommand extends SlashCommandAdapter {
     private final TagSystem tagSystem;
 
+    private static final String COMMAND_NAME = "tag";
     static final String ID_OPTION = "id";
     static final String REPLY_TO_USER_OPTION = "reply-to";
+    private static final int MAX_OPTIONS = 25;
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:31]: OptionData#MAX_OPTIONS or OptionData#MAX_CHOICES

Not sure what it's called

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 1 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:76]: yes, i figured that out already, will remove it

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 3 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:31]: MAX_CHOICES it is

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 2 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:80]: i don't get what u mean, can u elaborate

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: There's a StringDistance class or something, take a look at that class.

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: can't find any

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: Look again, StringDistances

IntelliJ has a feature which allows you to look for classes, you cannot miss it when using that

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: 2022-10-01_115204-751x578_scrot

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: In IntelliJ I said, maybe it's a class in our project (yes it is :p

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: ok found it

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: it won't work i believe it only returns a single item not a collection

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: startsWith should be just fine

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: Pretty sure I tagged you in the Discord about this.

Add an autocompletion method which does the same as autocomplete, but then it returns a collection

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:78]: It is "fine" but I much rather have better
See my messages in the Discord

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 4]
```diff
@@ -36,14 +44,16 @@ public TagCommand(TagSystem tagSystem) {
...
+                new OptionData(OptionType.STRING, ID_OPTION, "The id of the tag to display", true,
+                        true),
+                new OptionData(OptionType.USER, REPLY_TO_USER_OPTION,
+                        "Optionally, the user who you want to reply to", false));
     }
 
     @Override
     public void onSlashCommand(SlashCommandInteractionEvent event) {
-        String id = Objects.requireNonNull(event.getOption(ID_OPTION)).getAsString();
+        String id = event.getOption(ID_OPTION).getAsString();
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:56]: Removed require non null?

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 4 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:56]: yes, coz it can't be null

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:56]: Thumbsup

**Comment by Zabuzard** [comment]
Does this conflict with #602 ? Im confused.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 5]
```diff
@@ -62,4 +69,20 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
         }
         message.queue();
     }
+
+    @Override
+    public void onAutoComplete(CommandAutoCompleteInteractionEvent event) {
+        AutoCompleteQuery focusedOption = event.getFocusedOption();
+
+        if (focusedOption.getName().equals(ID_OPTION)) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:77]: 1. use early-return to reduce the nesting
2. use fail-fast and not defensive programming here. u dont expect other options to end up here, so if it does, it should throw

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 6]
```diff
@@ -51,6 +51,29 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
...
+            .min()
+            .orElse(0);
+        return candidates.stream()
+            .map(candidate -> new CandidateScore(candidate, prefixEditDistance(prefix, candidate)))
+            .sorted()
+            .takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore
+                    + firstCandidateScore * errorMargin)
+            .map(CandidateScore::candidate)
+            .toList();
+    }
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: the whole design for this method is not really "idiomatic". and there are multiple issues with it. lets refactor it a bit:
```java
public static Collection<String> closestMatches(CharSequence source, Collection<String> destinations, int limit) {
        // TODO some argument validation ...
        
        Queue<String> matches = new PriorityQueue<>(Comparator.comparingInt(x -> prefixEditDistance(source, x)));
        matches.putAll(destinations);
        
        return Stream.generate(matches::poll)
          .limit(limit)
          .toList();
}
```

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 7]
```diff
@@ -51,6 +51,29 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
...
 
+    /**
+     * Gives sorted suggestion to autocomplete a prefix string from the given options
+     *
+     * @param prefix the prefix to give suggestion for
+     * @param candidates all the possible suggestions
+     * @param errorMargin error margin of suggestions, error percentage / 100
+     * @return collection of autocomplete suggestions
+     */
+    public static Collection<String> autocompleteSuggestions(CharSequence prefix,
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:62]: Expand the existing unit tests for this. please dont leave it untested.

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/CandidateScore.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package org.togetherjava.tjbot.commands.utils;
+
+record CandidateScore(String candidate, int score) implements Comparable<CandidateScore> {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/CandidateScore.java:3]: This class is obsolete when u use the comparator-api (as shown in my rework example)

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 9]
```diff
@@ -62,4 +69,20 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
+            Collection<Command.Choice> choices = StringDistances
+                .autocompleteSuggestions(focusedOption.getValue(), tagSystem.getAllIds(), 0.5)
+                .stream()
+                .map(id -> new Command.Choice(id, id))
+                .limit(OptionData.MAX_CHOICES)
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:82]: this limit should be given to the `autocompleteSuggestions` method instead, it shouldnt sort ALL tags (which could theoretically be thousands). (see my suggested rework)

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 10]
```diff
@@ -62,4 +69,20 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
+                .toList();
+
+            event.replyChoices(choices).queue();
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:85]: `tagSystem.getAllIds()` can take a bit since it has to check the database. not sure how autocomplete works with long methods, but in regular commands u have to use `deferReply()` and then respond via the interactionhook. maybe u have to do sth like that here as well?

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 6 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: Can you elaborate on the issues, as it seems youre removing part of the functionality.

There's a reason it gets mapped to a CandidateScore, take a look at the takeWhile method.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 10 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:85]: Auto completion should be fast, I don't think there's deferReply to begin with, but even if there was, having to wait 3s for auto completion is way too long if you ask me.

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:85]: `CommandAutoCompleteInteractionEvent` have no method called `deferReply()` also i don't think `replyChoices()` timesout

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 6 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: The way this math attempts to provide a good UX isnt really working out imo. Its completely missing the string length for example.

I am okay with adding this feature again. But then incorperate it into the proposed design with the PQ instead. The `CandidateScore` class should be nested, nobody else needs it. And should be renamed to `MatchScore` or `ScoredMatch`, depending on what it contains.

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: I certainly believe this is far from the best solution.
but I do believe it's better than your suggestion of deleting it completely.

Since I personally lack experience with algorithms, you're on your own here.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: To begin with, i dont even understand how this math is supposed work at all. Its definitely in need for its own little method/comments/javadoc. And I would appreciate it if someone can give me a quick example calculation, so that i understand it.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: probably bc u made it public? hard to tell for me without seeing what u saw.

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:75]: should i put the record in the method?

**Comment by Taz03** [comment]
> Sth I just saw: This PR should remove the TODO at `TagCommand` line `42` that essentially asks about adding exactly this.

it says `if the user calls this without id or similar` and this pr never deals with that ig #602 does that

**Comment by Zabuzard** [comment]
@Taz03 that TODO was added bc we never knew that autocomplete will exist. the purpose of the TODO was to essentially provide autocomplete.

i know that bc i wrote it, lol 😆 its okay, please remove it, thanks

**Comment by Taz03** [comment]
ok

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 11]
```diff
@@ -177,4 +164,17 @@ private static int[][] computeLevenshteinDistanceTable(CharSequence source,
 
         return table;
     }
+
+    private record MatchScore(String candidate, int score) implements Comparable<MatchScore> {
+        @Override
+        public int compareTo(MatchScore otherMatchScore) {
+            int compare = Integer.compare(this.score, otherMatchScore.score);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:171]: `compare` is a bad name. if at all, then `compareResult`. that said, u can call it `byScore` or `byScoreResult` maybe

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 11 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:171]: what about scoreDiff?

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: `diff` suggests that it would be the difference, like `foo - bar`. But its not.

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 6 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: not resolved. still missing the priority queue approach with k-sort. and all the explanation/rework regarding that odd math.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 5 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:75]: not resolved. u didnt do this at all

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 7 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:62]: not resolved. u didnt do this at all

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 9 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java]: not resolved. u didnt do this at all

**Review by Zabuzard — CHANGES_REQUESTED** [review]
please resolve my existing issues, thanks

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 6 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: I thought u and tais agreed on keeping the current implementation, sry english is not my first language

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: no worries. we agreed on keeping a filtering to not display non-sense suggestions. but:
* this can still be added to the PQ k-sort approach perfectly fine
* and im not okay with this odd way of doing it and would like to get a small explanation of how it works to understand it and then suggest something better

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 7 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:62]: i did added some tests

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 6 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: The idea of this was made up by me, I don't know or there's any official algorithm like this.

The most confusing method is 
```java
            .takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore
                    + firstCandidateScore * errorMargin)
```

The idea of this method is pretty simple.

If the most accurate completion is 2 characters away, the most inaccurate completion can only be a max of 6 characters away. A slight improvement would do this.

```java
            .takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore * 2 * errorMargin)
```

Maybe a helper method would help even more?

```java
isCandidateInErrorRange(candidateScore , firstCandidateScore, errorMargin)
```

Do you understand it now?

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 7 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:62]: im sorry, i didnt see them.

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 6 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: sorry, no. can u give me a concrete example with some toy data?

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: We have input `autocomplet`

the most accurate completion is `autocomplete` which is a total of 1 character away

1 * 2 * 0.5 =  3
The farthest distance a completion can be away is 3, so any completion that has a stringdistance of `1/2/3` will be accepted.



Another completion, `autocom` is total of 4 character, this means this won't get accepted.
But `autocomp` will get accepted, since it's only 3 characters away.


This way, we won't just have 25 random completion.
Makes more sense?

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: yes. i understand the math now. it feels reversed to what i would have done:
```java
top10Candidates
  .filter(candidateWithPed -> candidateWithPed.ped / input.length()  < 0.5)
```
example, user inputs `"aaa"` and we have candidates `"aab"` and `"foo"`. latter is totally off, the PEDs are:
* `"aab" -> 1
* `"foo"` -> 3

the filter divides the PED with the original length `"aaa".length() -> 3` to compute how far away the candidate is:
* `"aab" -> `1 / 3` -> `0.333` is `< 0.5`, accepted
* `"foo"` -> `3 / 3` -> `1.0` is not `< 0.5`, discarded

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: I'm noticing some knowledge about algorithms, sounds better than my suggestion. \:p

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: this can ofc be optimized to already execute on the original stream with a `takeWhile`, like u suggested.

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]:
```java
.takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore * 2 * errorMargin)
```
@Tais993 this is not the same thing

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: what's the difference between `firstCandidateScore  * 2` and `firstCandidateScore + firstCandidateScore `

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: Actually, the way it was written (urs, but also mine) both have a performance flaw. We have to compute the PED for all entries once and then remember that instead. Using the PED in the Comparator will result in it being recomputed many times (more for sorted() than for a PQ, but still way too often). So we should do it like this:
```java
List<MatchScore> scoredMatches = getAllTags().stream()
  .map(match -> computePed(input, match))
  .toList();
```
and now we can sort based on the already-computed score:
```java
Queue<MatchScore> bestMatches = new PriorityQueue<>();
bestMatches.addAll(scoredMatches);

List<String> results = Stream.generate(bestMatches::poll)
  .takeWhile(match -> match.score / input.length() <= 0.5)
  .map(MatchScore::getMatch)
  .limit(10)
  .toList();
```
or similar (writing this from memory)

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: > what's the difference between `firstCandidateScore * 2` and `firstCandidateScore + firstCandidateScore `

thats not what it is, i first multiply errorMargin in firstCandidateScore and then add it to firstCandidateScore

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: Ah I completely forgot about mathematical rules LMFAO

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java]: Why not just do `firstCandidateScore * (errorMargin + 1)`, since this is identical

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 5 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:75]: which error to throw?

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 12]
```diff
@@ -1,15 +1,19 @@
...
 
 /**
  * Utility class for computing string distances, for example the edit distance between two words.
  */
 public class StringDistances {
+    /**
+     * Matches that are further off than this are not considered as match anymore. The value is
+     * between 0.0 (full match) and 1.0 (completely different).
+     */
+    private static final double offByPercentageThreshold = 0.5;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:15]: constant naming convention `OFF_BY_PERCENTAGE_THRESHOLD`

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 13]
```diff
@@ -56,24 +60,28 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
...
      * @return collection of autocomplete suggestions
      */
-    public static Collection<String> autocompleteSuggestions(CharSequence prefix,
-            Collection<String> candidates, double errorMargin) {
-        int firstCandidateScore = candidates.stream()
-            .mapToInt(candidate -> prefixEditDistance(prefix, candidate))
-            .min()
-            .orElse(0);
-        return candidates.stream()
+    public static Collection<String> closeMatches(CharSequence prefix,
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:65]: `limit` is missing javadoc

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 14]
```diff
@@ -56,24 +60,28 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
      *
      * @param prefix the prefix to give suggestion for
      * @param candidates all the possible suggestions
-     * @param errorMargin error margin of suggestions, error percentage / 100
      * @return collection of autocomplete suggestions
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:63]: `* @return collection of autocomplete suggestions` this text assumes context that is incorrect from the POV of `StringDistances`. rephrase to `the matches closest to the given prefix, limited to the given limit`

File: application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java [hunk 15]
```diff
@@ -10,25 +10,17 @@
...
-        List<TestCase> tests =
-                List.of(new TestCase("empty_candidates", List.of(), "prefix", List.of(), 0),
-                        new TestCase("empty_prefix", List.of("one", "three", "two"), "",
-                                List.of("one", "two", "three"), 0),
-                        new TestCase("all_empty", List.of(), "", List.of(), 1),
-                        new TestCase("max_error", List.of("aa"), "a",
-                                List.of("json", "one", "aa", "two", "++"), 1),
-                        new TestCase("real_test", List.of("j", "java", "js", "one"), "jo",
-                                List.of("java", "xj", "bs", "one", "yes", "js", "a", "j"), 0.5));
+        List<TestCase> tests = List.of();
```
Comment (Zabuzard) [inline application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java:18]: TODO tests

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java (hunk 5 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:75]: `IllegalArgumentException` since u got an option that u didnt expect.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 16]
```diff
@@ -72,15 +72,15 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
     public void onAutoComplete(CommandAutoCompleteInteractionEvent event) {
         AutoCompleteQuery focusedOption = event.getFocusedOption();
 
-        if (focusedOption.getName().equals(ID_OPTION)) {
-            return;
+        if (!focusedOption.getName().equals(ID_OPTION)) {
+            throw new IllegalArgumentException("Unexpected option");
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:76]: the exception would be a lot more helpful if u actually add what it was to the message. `"Unexpected option, was: " + ...`

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 17]
```diff
@@ -72,15 +72,15 @@ public void onSlashCommand(SlashCommandInteractionEvent event) {
...
         }
 
-        Collection<Command.Choice> choices = StringDistances
-            .closeMatches(focusedOption.getValue(), tagSystem.getAllIds(), OptionData.MAX_CHOICES)
-            .stream()
-            .map(id -> new Command.Choice(id, id))
-            .toList();
+        Collection<Command.Choice> choices =
+                StringDistances.closeMatches(focusedOption.getValue(), tagSystem.getAllIds(), 5)
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:80]: magic number `5`

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 18]
```diff
@@ -65,6 +65,14 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
      */
     public static Collection<String> closeMatches(CharSequence prefix,
             Collection<String> candidates, int limit) {
+        if (prefix.isEmpty()) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:68]: if the prefix is empty, the returned list should be empty, since any suggestion can only be a bad suggestion.

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 19]
```diff
@@ -65,6 +65,14 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
...
+            return candidates.stream().sorted().limit(limit).toList();
+        }
+
+        if (candidates.isEmpty()) {
+            return candidates;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:73]: personally, i prefer `return List.of();` in such a case. makes it more explicit and clear that the returned list is empty.

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 20]
```diff
@@ -73,14 +81,14 @@ public static Collection<String> closeMatches(CharSequence prefix,
...
+            .limit(limit)
             .takeWhile(matchScore -> isCloseEnough(matchScore, prefix))
             .map(MatchScore::candidate)
-            .limit(limit)
             .toList();
     }
 
     private static boolean isCloseEnough(MatchScore matchScore, CharSequence prefix) {
-        return matchScore.score / prefix.length() <= OFF_BY_PERCENTAGE_THRESHOLD;
+        return (double) matchScore.score / prefix.length() <= OFF_BY_PERCENTAGE_THRESHOLD;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:91]: voting for `score` itself to be `double` aready. based on the name, one would except that at least.

File: application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java [hunk 21]
```diff
@@ -15,7 +15,16 @@ record TestCase(String name, Collection<String> expectedSuggestions, String pref
                 Collection<String> candidates, int limit) {
         }
 
-        List<TestCase> tests = List.of();
+        List<String> exampleTags = List.of("c", "c#", "c++", "emacs", "foo", "hello", "java", "js",
```
Comment (Zabuzard) [inline application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java:18]: call it `candidates` maybe. or `existingTags`

File: application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java [hunk 22]
```diff
@@ -15,7 +15,16 @@ record TestCase(String name, Collection<String> expectedSuggestions, String pref
...
+                "key", "nvim", "py", "tag", "taz", "vi", "vim");
+
+        List<TestCase> tests = List.of(new TestCase("no_tags", List.of(), "foo", List.of(), 5),
+                new TestCase("no_prefix", List.of("c", "c#", "c++", "emacs", "foo"), "",
+                        exampleTags, 5),
+                new TestCase("both_empty", List.of(), "", List.of(), 5),
+                new TestCase("test0", List.of("vi", "vim"), "v", exampleTags, 5),
+                new TestCase("test1", List.of("java", "js"), "j", exampleTags, 5),
+                new TestCase("test2", List.of("c", "c#", "c++"), "c", exampleTags, 5));
```
Comment (Zabuzard) [inline application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java:27]: lots of hardcoded `5` here. add that as a variable beforehand

Comment (Zabuzard) [inline application/src/test/java/org/togetherjava/tjbot/commands/utils/StringDistancesTest.java:27]: `test0-2` arent really helpful names. maybe u can be a bit more precise as to what it tests. maybe `withPrefix0-2`

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 18 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:68]: yes, but i dont like to see the no auto-complete suggestions even before i did anything
image

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 20 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:91]: when PED method returns `int` it should be `int` ig but idk, i'll make it a `double` if u want

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 18 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:68]: lol whats that. cant it just tell discord that it has nothing and discord doesnt show anything extra then? like without autocomplete? @Tais993

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 20 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:91]: the PED should stay int but the score itself should be double. in the future, it might also include relevance or other stuff, BM-25 matching etc blabla. it doesnt necessarily have to be a PED only. otherwise, the name should be changed from score to distance imo - then it could stay an int 🤷

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java [hunk 23]
```diff
@@ -25,7 +25,7 @@
  */
 public final class TagCommand extends SlashCommandAdapter {
     private final TagSystem tagSystem;
-
+    private static final int MAX_OPTIONS = 5;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagCommand.java:28]: `MAX_OPTIONS` feels a bit confusing, naming-wise. i think it would be clearer if its `MAX_SUGGESTIONS` or `MAX_MATCHES`.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java (hunk 18 above)
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:68]: no

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:68]: that would be nice tho

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:68]: anyways, ur doing sorted() again instead of k-sort. thats the same mistake than before, a performance-problem. use a PQ.

File: application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java [hunk 24]
```diff
@@ -80,6 +76,13 @@ public static Collection<String> closeMatches(CharSequence prefix,
         Queue<MatchScore> bestMatches = new PriorityQueue<>();
         bestMatches.addAll(scoredMatches);
 
+        if (prefix.isEmpty()) {
+            return Stream.generate(bestMatches::poll)
+                .limit(limit)
+                .map(MatchScore::candidate)
+                .toList();
+        }
+
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java:85]: I feel like this feature is better put into `isCloseEnough` to always say `0.0` when the `prefix.length()` is `0`.

Right now, this feels like an useless duplication

**Review by Zabuzard — CHANGES_REQUESTED** [review]

**Review by Zabuzard — APPROVED** [review]
before u merge, please have a final test over all edge cases. its easy to break stuff through longer rounds of CR if it doesnt have 100% unit test coverage 👍

**Comment by Taz03** [comment]
@Zabuzard one approve and been a whole week so merge now?

**Comment by Tais993** [comment]
Okay so I learned something new.
If GitHub says "change requested" it's useless and means nothing, as my review has been requested already.
I'll review RN

**Review by Tais993 — APPROVED** [review]
